File: study1/solutions/week4/flood_fill.py
# https://leetcode.com/problems/flood-fill/discussion/

class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        if image[sr][sc] == color:
            return image
        to_traverse = [(sr, sc)]
        row_count = len(image)
        column_count = len(image[0])
        original_color = image[sr][sc]
        image[sr][sc] = color

        while to_traverse:
            row, column = to_traverse.pop()
            if row > 0 and image[row - 1][column] == original_color:
                to_traverse.append((row - 1, column))
                image[row-1][column] = color
            if column > 0 and image[row][column - 1] == original_color:
                to_traverse.append((row, column - 1))
                image[row][column-1] = color
            if row < row_count - 1 and image[row + 1][column] == original_color:
                to_traverse.append((row + 1, column))
                image[row+1][column] = color
            if column < column_count - 1 and image[row][column + 1] == original_color:
                to_traverse.append((row, column + 1))
                image[row][column+1] = color

        return image

File: study1/week4/flood-fill.java
// https://leetcode.com/problems/flood-fill/
// Time Complexity: O(n*m), n: image.length, m: image[0].length
// Space Complexity: O(1)
// using recursive methods
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        if (image[sr][sc] == color) {
            return image;
        }

        int pixelColor = image[sr][sc]; // original color

        image[sr][sc] = color; // change the color of the pixel

        int[][] dir = new int[][]{{0, 1},{0, -1},{1, 0},{-1, 0}}; // up, down, right, left
        for (int i = 0; i < dir.length; ++i) {
            int nr = sr + dir[i][0];
            int nc = sc + dir[i][1];

            if (nr < 0 || nr >= image.length || nc < 0 || nc >= image[0].length) {
                continue;
            }
            if (image[nr][nc] == pixelColor) {
                floodFill(image, nr, nc, color);
            }
        }

        return image;
    }
}